server/broadcast_test.go

```diff
@@ -338,3 +376,112 @@ func TestCleanupSessions(t *testing.T) {
 // Note: Add processSegment tests, including:
 //     assert an error from transcoder removes sess from BroadcastSessionManager
 //     assert a success re-adds sess to BroadcastSessionManager
+
+func TestTranscodeSegment_VerifyPixels(t *testing.T) {
+	require := require.New(t)
+	assert := assert.New(t)
+
+	dummyRes := func(tSegData []*net.TranscodedSegmentData) *net.TranscodeResult {
+		return &net.TranscodeResult{
+			Result: &net.TranscodeResult_Data{
+				Data: &net.TranscodeData{
+					Segments: tSegData,
+					Sig:      []byte("bar"),
+				},
+			},
+		}
+	}
+
+	// Create stub response with incorrect reported pixels
+	tSegData := []*net.TranscodedSegmentData{
+		&net.TranscodedSegmentData{Url: "test.flv", Pixels: 100},
+	}
+	tr := dummyRes(tSegData)
+	buf, err := proto.Marshal(tr)
+	require.Nil(err)
+
+	// Create stub server
+	ts, mux := stubTLSServer()
+	defer ts.Close()
+	mux.HandleFunc("/segment", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		w.Write(buf)
+	})
+
+	sess := StubBroadcastSession(ts.URL)
+	sess.Profiles = []ffmpeg.VideoProfile{ffmpeg.P144p30fps16x9}
+	bsm := bsmWithSessList([]*BroadcastSession{sess})
+	cxn := &rtmpConnection{
+		mid:         core.ManifestID("foo"),
+		nonce:       7,
+		pl:          &stubPlaylistManager{core.ManifestID("foo")},
+		profile:     &ffmpeg.P144p30fps16x9,
+		sessManager: bsm,
+	}
+
+	err = transcodeSegment(cxn, &stream.HLSSegment{Data: []byte("dummy")}, "dummy")
+	assert.Nil(err)
+
+	// Wait for async pixels verification to finish
+	time.Sleep(500 * time.Millisecond)
```

Comment:
I've been wondering if there's a better way to test against this type of pattern since sleeps tend to be racy, but can't think of anything that isn't unwieldy (eg, callback that's only used for testing).

Reply:
One thought is to make `cxn.sessManager` a interface instead of a concrete type. Then, during a test the stub for `cxn.sessManager` could implement `removeSession()` as:

```
type stubSessManager struct {
    removed chan *BroadcastSession
}

func (s *stubSessManager) removeSession(sess *BroadcastSession) {
    s.removed <- sess
}
```

In the test, we could do:

```
bsm := &stubSessManager{}

// Other test code

err := transcodeSeg(...)

sess := <-sessManager.removed

// Assert that sess is the BroadcastSession that we expect to be removed i.e. by checking
// sess.OrchestratorInfo.Transcoder
```

This would require a bit of refactoring though, so perhaps we can address the racy sleeps separately

Reply:
That sounds good. Seems like there's still the possibility of `removeSession` being called for a separate / unrelated failure, so we can't always be assured that would be testing what we want it to test